homeassistant/components/mysensors.py

```diff
@@ -0,0 +1,219 @@
+
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START,
+    EVENT_HOMEASSISTANT_STOP,
+    TEMP_CELCIUS)
+
+CONF_PORT = 'port'
+CONF_DEBUG = 'debug'
+CONF_PERSISTENCE = 'persistence'
+CONF_PERSISTENCE_FILE = 'persistence_file'
+CONF_VERSION = 'version'
+
+DOMAIN = 'mysensors'
+DEPENDENCIES = []
+REQUIREMENTS = [
+    'https://github.com/theolind/pymysensors/archive/'
+    '2aa8f32908e8c5bb3e5c77c5851db778f8635792.zip#pymysensors==0.3']
+_LOGGER = logging.getLogger(__name__)
+ATTR_PORT = 'port'
+ATTR_DEVICES = 'devices'
+ATTR_NODE_ID = 'node_id'
+ATTR_CHILD_ID = 'child_id'
+ATTR_UPDATE_TYPE = 'update_type'
+
+IS_METRIC = None
+CONST = None
+GATEWAYS = None
+EVENT_MYSENSORS_NODE_UPDATE = 'MYSENSORS_NODE_UPDATE'
+
+
+def setup(hass, config):  # noqa
+    """ Setup the MySensors component. """
+    # pylint:disable=no-name-in-module
+    import mysensors.mysensors as mysensors
+
+    if not validate_config(config,
+                           {DOMAIN: [CONF_PORT]},
+                           _LOGGER):
+        return False
+
+    version = config[DOMAIN].get(CONF_VERSION, '1.4')
+
+    global CONST
+    if version == '1.4':
+        import mysensors.const_14 as const
+        CONST = const
+    elif version == '1.5':
+        import mysensors.const_15 as const
+        CONST = const
+    else:
+        import mysensors.const_14 as const
+        CONST = const
+
+    # Just assume celcius means that the user wants metric for now.
+    # It may make more sense to make this a global config option in the future.
+    global IS_METRIC
+    IS_METRIC = (hass.config.temperature_unit == TEMP_CELCIUS)
+
+    def callback_generator(port, devices):
+        """Return a new callback function. Run once per gateway setup."""
+        def node_update(update_type, nid):
+            """Callback for node updates from the MySensors gateway."""
+            _LOGGER.info('update %s: node %s', update_type, nid)
+
+            hass.bus.fire(EVENT_MYSENSORS_NODE_UPDATE, {
+                ATTR_PORT: port,
+                ATTR_DEVICES: devices,
+                ATTR_UPDATE_TYPE: update_type,
+                ATTR_NODE_ID: nid
+            })
+            return
+        return node_update
+
+    def setup_gateway(port, persistence, persistence_file):
+        """Return gateway after setup of the gateway."""
+        devices = {}    # keep track of devices added to HA
+        gateway = mysensors.SerialGateway(port,
+                                          persistence=persistence,
+                                          persistence_file=persistence_file,
+                                          protocol_version=version)
+        gateway.event_callback = callback_generator(port, devices)
+        gateway.metric = IS_METRIC
+        gateway.debug = config[DOMAIN].get(CONF_DEBUG, False)
+        gateway.start()
+
+        def persistence_update(event):
+            """Callback to trigger update from persistence file."""
+            for _ in range(2):
```

Comment:
So you're saying first time is setup device second time update? Wouldn't
the trick be to update the device on setup?

On Sat, Dec 19, 2015, 15:52 Martin Hjelmare <notifications@github.com>
wrote:

> In homeassistant/components/mysensors.py
> <https://github.com/balloob/home-assistant/pull/771#discussion_r48097992>:
>
> > +
> > +    def setup_gateway(port, persistence, persistence_file):
> > +        """Return gateway after setup of the gateway."""
> > +        devices = {}    # keep track of devices added to HA
> > +        gateway = mysensors.SerialGateway(port,
> > +                                          persistence=persistence,
> > +                                          persistence_file=persistence_
>
> file,
> > +                                          protocol_version=version)
> > +        gateway.event_callback = callback_generator(port, devices)
> > +        gateway.metric = IS_METRIC
> > +        gateway.debug = config[DOMAIN].get(CONF_DEBUG, False)
> > +        gateway.start()
> > +
> > +        def persistence_update(event):
> > +            """Callback to trigger update from persistence file."""
> > +            for _ in range(2):
>
> The cause, is how the loops and if...elif logic is setup in the decorator
> mysensors_update. I didn't invest so much time, trying to find a better
> way, but I'd like to both add devices and update the values in one go, if
> possible. But at the same time you don't want to code it too complex, or
> nested. I think, that's why it was coded that way in the first place.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/771/files#r48097992>.
>